kempner_hpc_handbook/scalability/distributed_computing.md

```diff
@@ -52,8 +52,70 @@ Pytorch and Pytorch lightning support multiple communicatoin backends including
 
 
 ### Ray
-[TODO]
 
+[Ray](https://ray.io) is a distributed computation library for python that supports both single and multi-node execution. Ray is designed to be easy to use and provides a number of high-level abstractions for distributed computing, including a task-based API, a distributed actor API, and a distributed data API. Ray also includes a number of utilities for distributed machine learning, including a distributed data loader and a distributed hyperparameter search library.
+
+It serves as the backend for a number of tools in the modern AI/ML ecosystem, including [vLLM](https://github.com/vllm-project/vllm), [DCLM](https://github.com/mlfoundations/dclm) and others. 
+
+Ray clusters can be run on your laptop as well as on SLURM. The methods of starting the cluster are different, but once the cluster is started the python API is the same, enabling easy development of Ray applications for the cluster on your laptop.
+
+#### Using Ray on SLURM
+
+In order to use ray on SLURM, you will need to start a ray cluster and ensure that it is using the proper number of resources. The following script can be used to start a ray cluster on SLURM. This script will start a ray head on the first node in the allocation and then start ray workers on the rest of the nodes. The script will then start a python script that uses ray to do distributed computation. Note the use of teh SLURM environment variables. By default, Ray will attempt to use all CPUs on a node regardless of the number of tasks requested. This can be controlled by setting the `--num-cpus` flag when starting the ray cluster.
+
+```
+#! /bin/bash
+#SBATCH --job-name=ray_cluster
+#SBATCH -N 2 # number of nodes, this script is adaptable for any number of nodes
+#SBATCH --tasks-per-node=1
+#SBATCH --cpus-per-task=40 # number of cpus per task, this script can work with any number of cpus
+#SBATCH --gres=gpu:4 # number of gpus per node
+#SBATCH --mem=0 # memory per node, note that this is per node and not per task
+#SBATCH --time=0:30:00 # time limit for the job
+#SBATCH -p kempner_requeue
+#SBATCH --account=<your account>
+
+
+
+
+# Load environment with ray installed, as well as other dependencies
+# Can also run ray inside singularitiy containers, assuming that all needed dependencies are installed in the container
+module load python
+
+mamba activate fake_ray_environment
```

Comment:
Please provide info on how to create the `fake_ray_environment`

Reply:
Ray's a basic pip install, adding basic installation instructions though

Reply:
So that is an optional step? For users, who follow the handbook step by step, this line will be misleading.

Reply:
Semi-optional? Like you definitely don't want to install Ray in your base python implementation. I can maybe convert this just to a commend about "Placeholder for loading environment with Ray installed" or something more explicit like that

Reply:
Placeholder updated